reference/test/matrix/csr_kernels.cpp

```diff
@@ -1284,6 +1310,439 @@ TYPED_TEST(Csr, NonSquareMtxIsTransposable)
+        GKO_ASSERT_MTX_EQ_SPARSITY(permuted, this->mtx3_sorted);
+        ASSERT_TRUE(permuted->is_sorted_by_column_index());
+    }
+}
+
+
+TYPED_TEST(Csr, ScaledPermuteRectangular)
+{
+    using gko::matrix::permute_mode;
+    using value_type = typename TestFixture::value_type;
+
+    auto rpermuted =
+        this->mtx2->scale_permute(this->scale_perm2, permute_mode::rows);
+    auto irpermuted = this->mtx2->scale_permute(this->scale_perm2,
+                                                permute_mode::inverse_rows);
+    auto cpermuted =
+        this->mtx2->scale_permute(this->scale_perm3, permute_mode::columns);
+    auto icpermuted = this->mtx2->scale_permute(this->scale_perm3,
+                                                permute_mode::inverse_columns);
+    auto ref_rpermuted = ref_permute(this->mtx2.get(), this->scale_perm2.get(),
+                                     permute_mode::rows);
+    auto ref_irpermuted = ref_permute(this->mtx2.get(), this->scale_perm2.get(),
+                                      permute_mode::inverse_rows);
+    auto ref_cpermuted = ref_permute(this->mtx2.get(), this->scale_perm3.get(),
+                                     permute_mode::columns);
+    auto ref_icpermuted = ref_permute(this->mtx2.get(), this->scale_perm3.get(),
+                                      permute_mode::inverse_columns);
+
+    GKO_ASSERT_MTX_NEAR(rpermuted, ref_rpermuted, r<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(irpermuted, ref_irpermuted, r<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(cpermuted, ref_cpermuted, r<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(icpermuted, ref_icpermuted, r<value_type>::value);
+    GKO_ASSERT_MTX_EQ_SPARSITY(rpermuted, ref_rpermuted);
+    GKO_ASSERT_MTX_EQ_SPARSITY(irpermuted, ref_irpermuted);
+    GKO_ASSERT_MTX_EQ_SPARSITY(cpermuted, ref_cpermuted);
+    GKO_ASSERT_MTX_EQ_SPARSITY(icpermuted, ref_icpermuted);
+    ASSERT_TRUE(rpermuted->is_sorted_by_column_index());
+    ASSERT_TRUE(irpermuted->is_sorted_by_column_index());
+    ASSERT_TRUE(cpermuted->is_sorted_by_column_index());
+    ASSERT_TRUE(icpermuted->is_sorted_by_column_index());
+}
+
+
+TYPED_TEST(Csr, ScaledPermuteFailsWithIncorrectPermutationSize)
+{
+    using gko::matrix::permute_mode;
+
+    for (auto mode :
+         {/* no permute_mode::none */ permute_mode::rows, permute_mode::columns,
+          permute_mode::symmetric, permute_mode::inverse_rows,
+          permute_mode::inverse_columns, permute_mode::inverse_symmetric}) {
+        SCOPED_TRACE(mode);
+
+        ASSERT_THROW(this->mtx3_sorted->scale_permute(this->scale_perm0, mode),
+                     gko::ValueMismatch);
+    }
+}
+
+
+TYPED_TEST(Csr, NonsymmScaledPermute)
+{
+    using value_type = typename TestFixture::value_type;
+
+    auto permuted = this->mtx3_sorted->scale_permute(this->scale_perm3,
+                                                     this->scale_perm3_rev);
```

Comment:
Just to ensure my thought is correct, it is $PA{P^T}^T$ = $PAP$ so it is not the symmetric case